/* { dg-do preprocess } */
/* { dg-options "-std=c11 -pedantic" { target c } } */
/* { dg-options "-std=c++11 -pedantic" { target c++ } } */

Â¨ /* { dg-error "is not valid in an identifier" "" { target c++ } } */

BÌ€

Ì€ /* { dg-error "not valid at the start of an identifier" } */

AÌ€ /* { dg-warning "not in NFC" } */

ğ€€
ğŸ¿½	/* { dg-error "is not valid in an identifier" "" { target c++ } } */
ó¡ˆ´	/* { dg-error "is not valid in an identifier" "" { target c++ } } */
